### PR TITLE
Migrate encryption controllers to factory

### PR DESCRIPTION
**Pull request by mfojtik**
This migrates encryption controllers to use controller factory.

### CODE REVIEW

File: pkg/operator/encryption/controllers/controller.go [hunk 1]
```diff
@@ -28,9 +28,10 @@ func setUpInformers(
 
 	managedSecretsInformer := kubeInformersForNamespaces.InformersFor("openshift-config-managed").Core().V1().Secrets().Informer()
 	managedSecretsInformer.AddEventHandler(eventHandler)
+	deployer.AddEventHandler(eventHandler)
 
-	return append([]cache.InformerSynced{
+	return []cache.InformerSynced{
 		operatorInformer.HasSynced,
 		managedSecretsInformer.HasSynced,
```
Comment (sttts) [inline pkg/operator/encryption/controllers/controller.go]: deployer.HasSynched

File: pkg/operator/encryption/controllers/condition_controller.go [hunk 2]
```diff
@@ -54,26 +45,25 @@ func NewConditionController(
...
 	}
 
-	c.preRunCachesSynced = setUpInformers(deployer, operatorClient, kubeInformersForNamespaces, c.eventHandler())
-
-	return c
+	return factory.New().WithInformers(
+		kubeInformersForNamespaces.InformersFor("openshift-config-managed").Core().V1().Secrets().Informer(),
+		operatorClient.Informer(),
+		deployer,
+	).ResyncEvery(time.Second).WithSync(c.sync).ToController("EncryptionConditionController", eventRecorder.WithComponentSuffix("encryption-condition-controller"))
```
Comment (sttts) [inline pkg/operator/encryption/controllers/condition_controller.go]: resync every second?

File: pkg/controller/factory/factory.go [hunk 3]
```diff
@@ -16,14 +16,21 @@ type Factory struct {
...
-	informers             []cache.SharedInformer
+	informers             []RegistrableInformer
 	namespaceInformers    []*namespaceInformer
 	cachesToSync          []cache.InformerSynced
 	interestingNamespaces sets.String
 }
 
+// RegistrableInformer represents any structure that allow to register event handlers and informs if caches are synced.
+// Any SharedInformer will comply.
+type RegistrableInformer interface {
```
Comment (sttts) [inline pkg/controller/factory/factory.go]: this isn't part of the factory library?

Comment (mfojtik) [reply pkg/controller/factory/factory.go]: it is... it is a subset of SharedInformer interface so the "Deployer" thing can comply... We don't need full shared informer access... just event handler registration and has synced.

File: pkg/operator/encryption/controllers/controller.go (hunk 1 above)
Comment (mfojtik) [reply pkg/operator/encryption/controllers/controller.go]: good catch

File: pkg/operator/encryption/controllers/condition_controller.go (hunk 2 above)
Comment (mfojtik) [reply pkg/operator/encryption/controllers/condition_controller.go]: @sttts that was pre-existing https://github.com/openshift/library-go/pull/687/files/318a3bf613627b8148ed47788882d3f1a1490705#diff-a03dd41db6b9c24ab2e28afc2b14fe5dL199

Comment (sttts) [reply pkg/operator/encryption/controllers/condition_controller.go]: wow. Sounds like we would prefer something like 30 sec at least.

Comment (mfojtik) [reply pkg/operator/encryption/controllers/condition_controller.go]: @sttts this is all pre-existing... i would do this as a follow up maybe?

Comment (sttts) [reply pkg/operator/encryption/controllers/condition_controller.go]: follow-up is fine

File: pkg/operator/encryption/deployer/unionrevisionedpod.go [hunk 4]
```diff
@@ -13,9 +13,10 @@ import (
 // UnionRevisionLabelPodDeployer provides unified state from multiple distinct deployers.
 type UnionRevisionLabelPodDeployer struct {
 	delegates []statemachine.Deployer
+	hasSynced []cache.InformerSynced
 }
 
-var _ statemachine.Deployer = &RevisionLabelPodDeployer{}
+var _ statemachine.Deployer = &UnionRevisionLabelPodDeployer{}
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/unionrevisionedpod.go]: @p0lyn0mial @sttts this somehow slipped trough review.

**Comment by mfojtik** [comment]
unit tests are totally flaking

@stevekuznetsov @droslean have we done something in base images that might explain this?

File: pkg/operator/encryption/controllers/migration_controller.go [hunk 5]
```diff
@@ -58,12 +54,10 @@ const (
...
 type migrationController struct {
 	component string
+	name      string
 
 	operatorClient operatorv1helpers.OperatorClient
 
-	queue         workqueue.RateLimitingInterface
-	eventRecorder events.Recorder
-
 	preRunCachesSynced []cache.InformerSynced
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/controllers/migration_controller.go]: is `preRunCachesSynced` used ?

File: pkg/operator/encryption/controllers.go [hunk 6]
```diff
@@ -101,10 +103,10 @@ type Controllers struct {
 	controllers []runner
 }
 
-func (c *Controllers) Run(stopCh <-chan struct{}) {
+func (c *Controllers) Run(ctx context.Context, workers int) {
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/controllers.go]: just wondering if we want to expose `workers` IIRC all `encryption` controllers are single-threaded.

**Comment by droslean** [comment]
> unit tests are totally flaking

As far as I know, we didn't make any changes. Can someone debug this and make sure that this is an actual infrastructure problem?

**Pull request merged**